Create ZadachaA.c

smi97 opened the pull request.

smi97 committed:
Create ZadachaA.c

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+    while (1) {
+
+        if (!fgets (*buffer + buffer_move, buffer_size / buffer_count, stdin)) {
+            return -1;
```
в данном месте утечка памяти, утекает buffer

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+        if ((*buffer)[buffer_len + buffer_move - 1] != '\n') {
+            buffer_size *= 2;
+            buffer_count = 2;
+            *buffer = (char *) realloc (*buffer, buffer_size);
```
нельзя реаллочить память "саму в себя". следует делает realloc на какой-то временный указатель.
при таком коде:
`*buffer = (char *) realloc (*buffer, buffer_size);`
если realloc вернет NULL, то *buffer станет нулевым указателем, и его уже невозможно будет очистить при помощи free

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+#include <string.h>
+#include <stdlib.h>
+
+int read_one_string(char **buffer) {
```
непонятно, зачем самостоятельно писать чтение строки?
получается очень error prone код.
есть же, например, getline, который уже это умеет

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+    int strings_col = 100;
+    *strings = (char **) malloc (strings_col * sizeof (char *));
+    if (*strings == NULL) {
+        if (buffer) free (buffer);
```
неопределенное поведение.
поскольку buffer выше не был проинициализирован, то он указывает на случайную область памяти.
следовало выше определять буфер таким образом:
`char *buffer = NULL;`

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+        if (buffer) free (buffer);
+        return 0;
+    }
+    int size = 0, flag = 0;
```
size чего? flag чего?
следует называть переменные так, чтобы название отражало бы как-то их назначение

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+}
+
+int check_one(char *string) {
+    int iterator = 0, count = 0;
```
если область видимости переменной можно уменьшить, то следует это делать.
в данном случае, область видимости iterator - это только цикл for.

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+
+int check_brackets(char **strings_to_check, int size, char ***result_strings) {
+    int iterator = 0, result_size = 0;
+    *result_strings = (char **) malloc (size * sizeof (char *));
```
нет проверки на ошибку выделения памяти

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+
+
+int main() {
+    char **strings, **res;
```
неинициализированные указатели.

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+
+int main() {
+    char **strings, **res;
+    int size = read_all_strings (&strings), i = 0;
```
зачем так вообще писать, через запятую? это только добавляет сложности читателю кода.
объявляйте, инициализируйте переменные в разных строчка программы.

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,155 @@
+        free (strings);
+        free (res);
+        return 0;
+    } else {
```
можно не писать else. ведь в верхней векте кода уже стоит return

smi97 committed:
Made fixes according to comments

Removed my onestringreading, added getline
Added memory error checking
Some minor fixes

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,145 @@
+            if (*temp == NULL) {
+                return 0;
+            }
+            for(int i = 0; i < strings_amount; ++i) {
```
realloc сам скопирует старые указатели в новую память, так что удалять их не нужно

o2gy84 commented on ZadachaA.c:
```diff
@@ -0,0 +1,145 @@
+                return 0;
+            }
+        }
+        getline_result = getline(&(*strings)[strings_amount], &string_size, stdin);
```
еще прикол функции getline() в том, что ей можно передать неаллоцированную строку, и getline() сам внутри себя выделит столько памяти, сколько нужно
